libraries/chain/db_market.cpp

```diff
@@ -191,27 +191,101 @@ void database::cancel_order(const force_settlement_object& order, bool create_vi
+            core_cancel_fee.amount = deferred_fee;
+         // if there is any CORE fee to deduct, redirect it to referral program
+         if( core_cancel_fee.amount > 0 )
+         {
+            seller_acc_stats = &order.seller( *this ).statistics( *this );
+            modify( *seller_acc_stats, [&]( account_statistics_object& obj ) {
+               obj.pay_fee( core_cancel_fee.amount, get_global_properties().parameters.cashback_vesting_threshold );
+            } );
+            deferred_fee -= core_cancel_fee.amount;
+            // handle originally paid fee if any:
+            //    to_deduct = round_up( paid_fee * core_cancel_fee / deferred_core_fee_before_deduct )
+            if( deferred_paid_fee.amount == 0 )
+            {
+               vop.fee = core_cancel_fee;
+            }
+            else
+            {
+               fc::uint128 fee128( deferred_paid_fee.amount.value );
+               fee128 *= core_cancel_fee.amount.value;
+               // to round up
+               fee128 += order.deferred_fee.value;
+               fee128 -= 1;
+               fee128 /= order.deferred_fee.value;
+               share_type cancel_fee_amount = fee128.to_uint64();
+               // cancel_fee should be positive, pay it to asset's accumulated_fees
+               fee_asset_dyn_data = &deferred_paid_fee.asset_id(*this).dynamic_asset_data_id(*this);
+               modify( *fee_asset_dyn_data, [&](asset_dynamic_data_object& addo) {
+                  addo.accumulated_fees += cancel_fee_amount;
+               });
+               // cancel_fee should be no more than deferred_paid_fee
+               deferred_paid_fee.amount -= cancel_fee_amount;
+               vop.fee = asset( cancel_fee_amount, deferred_paid_fee.asset_id );
+            }
+         }
       }
-   });
+   }
+
+   // refund funds in order
+   auto refunded = order.amount_for_sale();
+   if( refunded.asset_id == asset_id_type() )
+   {
+      if( seller_acc_stats == nullptr )
+         seller_acc_stats = &order.seller( *this ).statistics( *this );
+      modify( *seller_acc_stats, [&]( account_statistics_object& obj ) {
+         obj.total_core_in_orders -= refunded.amount;
+      });
+   }
    adjust_balance(order.seller, refunded);
-   adjust_balance(order.seller, order.deferred_fee);
 
-   if( create_virtual_op )
+   // refund fee
+   // could be virtual op or real op here
+   // TODO check if skip_fee is enabled?
+   if( deferred_paid_fee.amount == 0 )
    {
-      limit_order_cancel_operation vop;
-      vop.order = order.id;
-      vop.fee_paying_account = order.seller;
-      push_applied_operation( vop );
+      // be here, order.create_time <= HARDFORK_CORE_604_TIME, or fee paid in CORE, or no fee to refund.
+      // if order was created before hard fork 604 then cancelled no matter before or after hard fork 604,
+      //    see it as fee paid in CORE, deferred_fee should be refunded to order owner but not fee pool
```

Comment:
It is possible that deferred_paid_fee.amount == 0 even if the fee was originally paid in non-CORE and deferred_fee is still > 0.
For example, if the original fee was paid in 1 satoshi of WHALESHARES, then the cancel_fee will be rounded up to 1 satoshi of WHALESHARES leaving deferred_paid_fee empty. deferred_fee will still contain the BTS-equivalent of 1 satoshi of WHALESHARES, minus the nominal cancel_fee.
It is correct to refund the remaining deferred_fee to the owner of the limit_order, which is what the code does. However, that leaves the possibility open to drain the pool of WHALESHARES. (The only other option is probably to consume the full deferred_fee for the cancel_operation.)

Reply:
Ok, makes sense.